Source/WebCore/css/StyleProperties.cpp

```diff
@@ -581,180 +581,211 @@ String StyleProperties::fontSynthesisValue() const
                 }
             }
+            skipSerializing = skipSerializing || !layerValue || isInitialValueForLonghand(longhand, *layerValue);
+            layerValues[j] = WTFMove(layerValue);
+            skipSerializingLayerValues[j] = skipSerializing;
+        }
 
-            auto canOmitValue = [&]() {
-                if (shorthand.id() == CSSPropertyMask) {
-                    if (property == CSSPropertyMaskClip) {
-                        // If the mask-clip value is the same as the value for mask-origin (the previous value),
-                        // then we can skip serializing it, as one value sets both properties.
-                        ASSERT(j > 0);
-                        ASSERT(shorthand.properties()[j - 1] == CSSPropertyMaskOrigin);
-                        auto originValue = values[j - 1];
-                        if (is<CSSValueList>(*originValue))
-                            originValue = downcast<CSSValueList>(*originValue).item(i);
-                        if (!is<CSSPrimitiveValue>(*value) || (originValue && !is<CSSPrimitiveValue>(*originValue)))
-                            return false;
-
-                        auto maskId = downcast<CSSPrimitiveValue>(*value).valueID();
-                        auto originId = originValue ? downcast<CSSPrimitiveValue>(*originValue).valueID() : CSSValueInitial;
-                        return maskId == originId && (!isCSSWideValueKeyword(StringView { nameLiteral(maskId) }) || value->isImplicitInitialValue());
-                    }
-                    if (property == CSSPropertyMaskOrigin) {
-                        // We can skip serializing mask-origin if it's the initial value, but only if we're also going to skip serializing
-                        // the mask-clip as well (otherwise the single value for mask-clip would be assumed to be setting the value for both).
-                        ASSERT(j + 1 < size);
-                        ASSERT(shorthand.properties()[j + 1] == CSSPropertyMaskClip);
-                        auto clipValue = values[j + 1];
-                        if (is<CSSValueList>(*clipValue))
-                            clipValue = downcast<CSSValueList>(*clipValue).item(i);
-                        return value->isImplicitInitialValue() && (!clipValue || clipValue->isImplicitInitialValue());
+        for (unsigned j = 0; j < length; j++) {
+            auto longhand = shorthand.properties()[j];
+
+            // A single box value sets both background-origin and background-clip.
+            // A single geometry-box value sets both mask-origin and mask-clip.
+            // A single geometry-box value sets both mask-origin and -webkit-mask-clip.
+            if (longhand == CSSPropertyBackgroundClip || longhand == CSSPropertyMaskClip || longhand == CSSPropertyWebkitMaskClip) {
+                // The previous property is origin.
+                auto areEqual = [](const RefPtr<CSSValue>& a, const RefPtr<CSSValue>& b) {
```

Comment:
Is this different than compareCSSValuePtr?

Reply:
It is. That function treats two null pointers as equal. But I can use it here I think.